src/test/java/com/amazon/ion/impl/bin/PresenceBitmapTest.kt

```diff
@@ -0,0 +1,330 @@
+    }
+
+    @Test
+    fun `when calling get with an invalid index, should throw exception`() {
+        val pb = PresenceBitmap()
+        val signature = listOf(taggedZeroOrOne, taggedOneToMany, taggedExactlyOne)
+        pb.initialize(signature)
+        assertThrows<IndexOutOfBoundsException> { pb.get(-1) }
+        assertThrows<IndexOutOfBoundsException> { pb.get(3) }
+    }
+
+    @Test
+    fun `when calling set, the presence bits value for that parameter is _not_ validated`() {
+        val signature = listOf(taggedZeroOrOne, taggedOneToMany, taggedExactlyOne)
+        with(PresenceBitmap()) {
+            initialize(signature)
+            // PresenceBits is an internal only class, so we rely on callers to do the correct thing.
+            // There should not be an exception thrown for any of these.
+            set(0, value = PresenceBitmap.GROUP)
+            set(1, value = PresenceBitmap.VOID)
+            set(2, value = PresenceBitmap.GROUP)
+            set(2, value = PresenceBitmap.VOID)
+        }
+    }
+
+    @ParameterizedTest
+    @CsvSource(
+        // For some reason `Long.decode()` doesn't support binary, so
+        // we're just using decimal for the presence values here.
+        "One, 0, false",
+        "One, 1, true",
+        "One, 2, false",
+        "One, 3, false",
+        "Any, 0, true",
+        "Any, 1, true",
+        "Any, 2, true",
+        "Any, 3, false",
+        "AtMostOne, 0, true",
+        "AtMostOne, 1, true",
+        "AtMostOne, 2, false",
+        "AtMostOne, 3, false",
+        "AtLeastOne, 0, false",
+        "AtLeastOne, 1, true",
+        "AtLeastOne, 2, true",
+        "AtLeastOne, 3, false",
+    )
+    fun `validate() correctly throws exception when presence bits are invalid for signature`(cardinality: ParameterCardinality, presenceValue: Long, isValid: Boolean) {
+        val signature = listOf(Parameter("a", ParameterEncoding.uint8, cardinality))
+        with(PresenceBitmap()) {
+            initialize(signature)
+            set(0, presenceValue)
+            if (isValid) {
+                validate()
+            } else {
+                assertThrows<IonException> { validate() }
+            }
+        }
+    }
+
+    @Test
+    fun `when all parameters are tagged and exactly-one, no presence bits are needed or written`() {
+        (0..128).forEach { n -> assertExpectedPresenceBitSizes(expectedByteSize = 0, signature = List(n) { taggedExactlyOne }) }
+    }
+
+    @Test
+    fun `when all parameters are tagless and exactly-one, no presence bits are needed or written`() {
+        (0..128).forEach { n -> assertExpectedPresenceBitSizes(expectedByteSize = 0, signature = List(n) { taglessExactlyOne }) }
+    }
+
+    @Test
+    fun `when all parameters are tagged and not exactly-one, should write expected number of presence bits`() {
+        // Index of an element in this list is the number of parameters in the signature
+        listOf(0, 0, 0, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3)
```

Comment:
Let's discuss the "2 or more tagged parameters" heuristic in an upcoming sync. We need to be clear how this would be encoded without presence bits.